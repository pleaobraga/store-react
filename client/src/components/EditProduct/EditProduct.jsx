import React from 'react'
import PropTypes from 'prop-types'
import { useHistory } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { putProduct, postProduct } from '../../reducers/ProductReducer'

import { formatReal } from '../../utils/utils'
import Product from '../Product'
import Loading from '../Loading'

import './EditProduct.scss'

export const EditProduct = ({ currentProduct, isEditing, storeName }) => {
  const dispatch = useDispatch()
  const pageContent = useSelector(state => state.product)
  const [product, setProduct] = React.useState({})
  const history = useHistory()

  React.useEffect(() => {
    setProduct({ ...currentProduct })
  }, [])

  const onChange = event => {
    if (event.target.name === 'price') {
      const price = formatReal(event.target.value)

      setProduct({ ...product, price })
    } else {
      setProduct({ ...product, [event.target.name]: event.target.value })
    }
  }

  const addErrorClass = className => {
    document
      .getElementsByClassName(className)[0]
      .parentElement.classList.add('error')
  }

  const removeErrorClass = () => {
    const allErrorsElements = document.querySelectorAll('.error')

    allErrorsElements.forEach(element => {
      element.classList.remove('error')
    })
  }

  const isFormValid = () => {
    let formValid = true

    if (!product.name || product.name.length <= 0) {
      addErrorClass('name')
      formValid = false
    }

    if (
      !product.quantity ||
      product.quantity.length <= 0 ||
      product.quantity < 0
    ) {
      addErrorClass('quantity')
      formValid = false
    }

    if (!product.currency || product.currency.length <= 0) {
      addErrorClass('currency')
      formValid = false
    }

    if (!product.price || product.price.length <= 0) {
      addErrorClass('price')
      formValid = false
    }

    return formValid
  }

  const save = event => {
    event.preventDefault()

    if (isFormValid()) {
      removeErrorClass()

      if (isEditing) {
        dispatch(putProduct(product))
      } else {
        dispatch(postProduct({ product, storeName }))
      }
    }
  }

  const cancel = event => {
    event.preventDefault()

    history.push(`/${storeName}`)
  }

  return (
    <form className="edit-product">
      <div className="edit-product__message">
        {pageContent.registeredError && (
          <p className="edit-product__message--error">
            This Product is already registered
          </p>
        )}
        {pageContent.updateProduct && (
          <p className="edit-product__message--saved">
            This Product was updated successfully
          </p>
        )}
      </div>

      <div className="edit-product__values">
        <div className="input-group">
          <label>Name</label>
          <input
            className="name"
            name="name"
            value={product.name}
            onChange={onChange}
            placeholder="Type the product name"
            required
          />
        </div>
        <div className="input-group">
          <label>Quantity</label>
          <input
            className="quantity"
            name="quantity"
            value={product.quantity}
            onChange={onChange}
            type="number"
            placeholder="qty"
            required
          />
        </div>
        <div className="input-group">
          <label>Currency</label>
          <input
            className="currency"
            name="currency"
            value={product.currency}
            onChange={onChange}
            placeholder="$"
            required
          />
        </div>

        <div className="input-group">
          <label>Price</label>
          <input
            className="price"
            name="price"
            value={product.price}
            onChange={onChange}
            required
          />
        </div>
      </div>
      <div className="edit-product__actions">
        <button className="save" type="submit" onClick={save}>
          {pageContent.loadingActions ? <Loading color="#fff" /> : 'Save'}
        </button>
        <button className="cancel" onClick={cancel}>
          Cancel
        </button>
      </div>
    </form>
  )
}

EditProduct.propTypes = {
  currentProduct: PropTypes.shape(Product.Prototype),
  isEditing: PropTypes.bool.isRequired,
  storeName: PropTypes.string
}

export default EditProduct
