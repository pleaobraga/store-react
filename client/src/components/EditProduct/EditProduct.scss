@import '../../styles/colors';

.edit-product {
  display: flex;
  flex-direction: column;

  &__message {
    * {
      font-size: 2rem;
      font-weight: 500;
      text-align: center;
      padding: 1rem 0;
    }

    &--error {
      color: $red;
      background-color: $light-pink;
    }

    &--saved {
      color: $green;
      background-color: $light-green;
    }
  }

  &__values {
    display: flex;
    flex-wrap: wrap;

    .input-group {
      margin: 1rem 3rem 1rem;

      &.error {
        label {
          color: $red;
        }

        input {
          border-color: $red;
          color: $red;
        }
      }

      label {
        font-size: 2rem;
        font-weight: 500;
        margin-right: 2rem;
      }

      input {
        font-size: 1.5rem;

        border-top: none;
        border-right: none;
        border-left: none;
        border-bottom: 1px solid $black;

        width: 10rem;

        &.name {
          width: 25rem;
        }

        &.quantity {
          width: 5rem;
        }

        &.currency {
          width: 3rem;
        }

        &.quantity {
          width: 5rem;
        }
      }
    }
  }

  &__actions {
    display: flex;
    justify-content: flex-end;

    button {
      margin: 2.5rem 1rem 1rem 1rem;
      padding: 0.5rem 1rem;

      cursor: pointer;

      border: 1px solid black;
      border-radius: 2px;

      box-shadow: 0px 3px 1px -2px rgba(0, 0, 0, 0.2),
        0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 1px 5px 0px rgba(0, 0, 0, 0.12);

      color: white;
      font-weight: 500;
      font-size: 1.5rem;

      line-height: 1.75;
      border-radius: 4px;
      letter-spacing: 0.02857em;
      text-transform: uppercase;

      &.save {
        background-color: $blue;
        border: 1px solid $blue;

        height: 4rem;
        width: 7rem;
      }

      &.cancel {
        background-color: $pink;
        border: 1px solid $pink;
      }

      &:focus {
        outline: none;
      }

      &:active {
        box-shadow: 0px 5px 5px -3px rgba(0, 0, 0, 0.2),
          0px 8px 10px 1px rgba(0, 0, 0, 0.14),
          0px 3px 14px 2px rgba(0, 0, 0, 0.12);
      }
    }
  }
}
