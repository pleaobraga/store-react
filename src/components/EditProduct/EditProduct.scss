@import '../../styles/colors';

.edit-product {
  display: flex;
  flex-direction: column;

  &__values {
    display: flex;
    flex-wrap: wrap;

    .input-group {
      margin: 1rem 3rem 1rem;

      label {
        font-size: 2rem;
        font-weight: 500;
        margin-right: 2rem;
      }

      input {
        border-top: none;
        border-right: none;
        border-left: none;
        border-bottom: 1px solid $black;

        width: 10rem;

        &.name {
          width: 25rem;
        }

        &.quantity {
          width: 5rem;
        }

        &.currency {
          width: 3rem;
        }

        &.quantity {
          width: 5rem;
        }
      }
    }
  }

  &__actions {
    display: flex;
    justify-content: flex-end;

    button {
      margin: 1rem;
      padding: 0.5rem;

      cursor: pointer;

      border: 1px solid black;
      border-radius: 2px;

      box-shadow: 0px 3px 1px -2px rgba(0, 0, 0, 0.2),
        0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 1px 5px 0px rgba(0, 0, 0, 0.12);

      color: white;
      font-weight: 500;
      font-size: 1.5rem;

      line-height: 1.75;
      border-radius: 4px;
      letter-spacing: 0.02857em;
      text-transform: uppercase;

      &.save {
        background-color: #1976d2;
        border: 1px solid #1976d2;
      }

      &.cancel {
        background-color: rgb(220, 0, 78);
        border: 1px solid rgb(220, 0, 78);
      }

      &:focus {
        outline: none;
      }

      &:active {
        box-shadow: 0px 5px 5px -3px rgba(0, 0, 0, 0.2),
          0px 8px 10px 1px rgba(0, 0, 0, 0.14),
          0px 3px 14px 2px rgba(0, 0, 0, 0.12);
      }
    }
  }
}
