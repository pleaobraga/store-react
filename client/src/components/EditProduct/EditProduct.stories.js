import React from 'react'
import { storiesOf } from '@storybook/react'
import { BrowserRouter as Router } from 'react-router-dom'
import ProviderWrapper from '../../../.storybook/provider'
import store from '../../../.storybook/configureStore'

import {
  initialStateRootReducer,
  product
} from '../../reducers/__mocks__/reduxMock'
import { EditProduct } from './EditProduct'

const newStore = store(initialStateRootReducer)

const withProvider = story => (
  <ProviderWrapper store={newStore}>{story()}</ProviderWrapper>
)

export default storiesOf('Components | EditProduct', module)
  .addDecorator(withProvider)
  .add(
    'default',
    () => (
      <Router>
        <EditProduct currentProduct={product} storeName="0" isEditing={true} />
      </Router>
    ),
    {
      info: {
        inline: true,
        header: false,
        propTables: [EditProduct]
      }
    }
  )
