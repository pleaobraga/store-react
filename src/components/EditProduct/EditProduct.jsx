import React from 'react'
import PropTypes from 'prop-types'
import { formatReal } from '../../utils/utils'
import './EditProduct.scss'

export const EditProduct = ({ name, quantity, price, currency }) => {
  const [product, setProduct] = React.useState({})

  React.useEffect(() => {
    setProduct({ name, quantity, price, currency })
  }, [])

  const onChange = event => {
    if (event.target.name === 'price') {
      const price = formatReal(event.target.value)

      setProduct({ ...product, price })
    } else {
      setProduct({ ...product, [event.target.name]: event.target.value })
    }
  }

  const save = event => {
    event.preventDefault()
  }

  const cancel = event => {
    event.preventDefault()
  }

  return (
    <form className="edit-product">
      <div className="edit-product__values">
        <div className="input-group">
          <label>Name</label>
          <input
            className="name"
            name="name"
            value={product.name}
            onChange={onChange}
            placeholder="name"
            required
          />
        </div>
        <div className="input-group">
          <label>Quantity</label>
          <input
            className="quantity"
            name="quantity"
            value={product.quantity}
            onChange={onChange}
            type="number"
            placeholder="qty"
            required
          />
        </div>
        <div className="input-group">
          <label>Currency</label>
          <input
            className="currency"
            name="currency"
            value={product.currency}
            onChange={onChange}
            placeholder="$"
            required
          />
        </div>

        <div className="input-group">
          <label>Price</label>
          <input
            className="price"
            name="price"
            value={product.price}
            onChange={onChange}
            required
          />
        </div>
      </div>
      <div className="edit-product__actions">
        <button className="save" onClick={save}>
          Save
        </button>
        <button className="cancel" onClick={cancel}>
          Cancel
        </button>
      </div>
    </form>
  )
}

EditProduct.defaultProps = {
  name: '',
  quantity: 1,
  price: '0,00',
  currency: 'R$'
}

EditProduct.propTypes = {
  name: PropTypes.string,
  quantity: PropTypes.number,
  price: PropTypes.string,
  currency: PropTypes.string
}

export default React.memo(EditProduct)
