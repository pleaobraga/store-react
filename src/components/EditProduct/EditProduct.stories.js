import React from 'react'
import { storiesOf } from '@storybook/react'
import { text, number } from '@storybook/addon-knobs'
import { BrowserRouter as Router } from 'react-router-dom'
import { Product } from './EditProduct'

export default storiesOf('Components | Product', module).add(
  'default',
  () => (
    <Router>
      <Product
        name={text('Name', 'Name')}
        quantity={number('Quantity', 2)}
        price={text('Price', '20,10')}
        currency={text('Currency', 'R$')}
      />
    </Router>
  ),
  {
    info: {
      inline: true,
      header: false,
      propTables: [Product]
    }
  }
)
