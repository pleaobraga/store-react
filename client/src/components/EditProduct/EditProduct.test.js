import React from 'react'
import { mount } from 'enzyme'
import { BrowserRouter as Router } from 'react-router-dom'
import { Provider } from 'react-redux'

import EditProduct from './EditProduct'
import {
  mockStore,
  initialStateRootReducer,
  product
} from '../../reducers/__mocks__/reduxMock'

const setup = ({ isEditing }) => {
  const store = mockStore(initialStateRootReducer)

  const contentPage = mount(
    <Provider store={store}>
      <Router>
        <EditProduct
          currentProduct={product}
          storeName="0"
          isEditing={isEditing}
        />
      </Router>
    </Provider>
  )

  return contentPage
}

describe('Product', () => {
  describe('Edit', () => {
    const product = setup({ isEditing: true })
    it('Should render proprely', () => {
      expect(product).toMatchSnapshot()
    })
  })

  describe('New Product', () => {
    const product = setup({ isEditing: false })
    it('should render proprely', () => {
      expect(product).toMatchSnapshot()
    })
  })
})
