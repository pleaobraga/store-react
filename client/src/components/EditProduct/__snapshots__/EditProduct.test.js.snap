// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Edit Should render proprely 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <EditProduct
        currentProduct={
          Object {
            "currency": "R$",
            "name": "Name",
            "price": "20,10",
            "quantity": 2,
          }
        }
        isEditing={true}
        storeName="0"
      >
        <form
          className="edit-product"
        >
          <div
            className="edit-product__message"
          />
          <div
            className="edit-product__values"
          >
            <div
              className="input-group"
            >
              <label>
                Name
              </label>
              <input
                className="name"
                name="name"
                onChange={[Function]}
                placeholder="Type the product name"
                required={true}
                value="Name"
              />
            </div>
            <div
              className="input-group"
            >
              <label>
                Quantity
              </label>
              <input
                className="quantity"
                name="quantity"
                onChange={[Function]}
                placeholder="qty"
                required={true}
                type="number"
                value={2}
              />
            </div>
            <div
              className="input-group"
            >
              <label>
                Currency
              </label>
              <input
                className="currency"
                name="currency"
                onChange={[Function]}
                placeholder="$"
                required={true}
                value="R$"
              />
            </div>
            <div
              className="input-group"
            >
              <label>
                Price
              </label>
              <input
                className="price"
                name="price"
                onChange={[Function]}
                required={true}
                value="20,10"
              />
            </div>
          </div>
          <div
            className="edit-product__actions"
          >
            <button
              className="save"
              onClick={[Function]}
              type="submit"
            >
              Save
            </button>
            <button
              className="cancel"
              onClick={[Function]}
            >
              Cancel
            </button>
          </div>
        </form>
      </EditProduct>
    </Router>
  </BrowserRouter>
</Provider>
`;

exports[`Product New Product should render proprely 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <EditProduct
        currentProduct={
          Object {
            "currency": "R$",
            "name": "Name",
            "price": "20,10",
            "quantity": 2,
          }
        }
        isEditing={false}
        storeName="0"
      >
        <form
          className="edit-product"
        >
          <div
            className="edit-product__message"
          />
          <div
            className="edit-product__values"
          >
            <div
              className="input-group"
            >
              <label>
                Name
              </label>
              <input
                className="name"
                name="name"
                onChange={[Function]}
                placeholder="Type the product name"
                required={true}
                value="Name"
              />
            </div>
            <div
              className="input-group"
            >
              <label>
                Quantity
              </label>
              <input
                className="quantity"
                name="quantity"
                onChange={[Function]}
                placeholder="qty"
                required={true}
                type="number"
                value={2}
              />
            </div>
            <div
              className="input-group"
            >
              <label>
                Currency
              </label>
              <input
                className="currency"
                name="currency"
                onChange={[Function]}
                placeholder="$"
                required={true}
                value="R$"
              />
            </div>
            <div
              className="input-group"
            >
              <label>
                Price
              </label>
              <input
                className="price"
                name="price"
                onChange={[Function]}
                required={true}
                value="20,10"
              />
            </div>
          </div>
          <div
            className="edit-product__actions"
          >
            <button
              className="save"
              onClick={[Function]}
              type="submit"
            >
              Save
            </button>
            <button
              className="cancel"
              onClick={[Function]}
            >
              Cancel
            </button>
          </div>
        </form>
      </EditProduct>
    </Router>
  </BrowserRouter>
</Provider>
`;
